spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,308 @@
+package org.apache.spark.sql.cassandra
+
+import scala.reflect.ClassTag
+
+import org.apache.spark.Logging
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.cassandra.CassandraSQLRow.CassandraSQLRowReader
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{types, DataFrame, Row, SQLContext}
+
+import com.datastax.spark.connector._
+import com.datastax.spark.connector.cql.{CassandraConnector, ColumnDef}
+import com.datastax.spark.connector.rdd.{ReadConf, ValidRDDType}
+import com.datastax.spark.connector.writer.{WriteConf, SqlRowWriter}
+
+import DataTypeConverter._
+
+/**
+ *  Implements [[BaseRelation]] and [[InsertableRelation]].
+ *  It inserts data into Cassandra table. It also provides some
+ *  helper methods and variables for sub class to use.
+ */
+private[cassandra] abstract class BaseRelationImpl(
+    tableIdent: TableIdent,
+    connector: CassandraConnector,
+    readConf: ReadConf,
+    writeConf: WriteConf,
+    userSpecifiedSchema: Option[StructType],
+    override val sqlContext: SQLContext)
+  extends BaseRelation
+  with InsertableRelation
+  with Serializable
+  with Logging {
+
+  protected[this] val tableDef =
+    sqlContext.getCassandraSchema(tableIdent.cluster.getOrElse("default"))
+      .keyspaceByName(tableIdent.keyspace).tableByName(tableIdent.table)
+
+  override def schema: StructType = {
+    userSpecifiedSchema.getOrElse(StructType(tableDef.allColumns.map(toStructField)))
+  }
+
+  protected[this] val baseRdd =
+    sqlContext.sparkContext.cassandraTable[CassandraSQLRow](
+      tableIdent.keyspace,
+      tableIdent.table)(
+        connector,
+        readConf,
+        implicitly[ClassTag[CassandraSQLRow]],
+        CassandraSQLRowReader,
+        implicitly[ValidRDDType[CassandraSQLRow]])
```

Comment:
There is no need to specify all those implicit  parameter values here. It is enough to define an `implicit val` for `connector` and `readConf` immediately before the `cassandraTable` call.